spec/spec.md

```diff
@@ -0,0 +1,2604 @@
+reference="last_connect"}), the conditional connects to a sub-element of
+an aggregate component only generates a multiplexer for the sub-element
+that is overwritten.
+
+For example, the following snippet:
+
+    wire x: {a:UInt, b:UInt}
+    wire y: UInt
+    wire c: UInt<1>
+    wire w: {a:UInt, b:UInt}
+    w <= x
+    when c :
+       w.a <= y
+    ...
+
+can be rewritten equivalently as follows:
+
+    wire x: {a:UInt, b:UInt}
+    wire y: UInt
+    wire c: UInt<1>
+    wire w: {a:UInt, b:UInt}
+    w.a <= mux(c, y, x.a)
+    w.b <= x.b
+    ...
+
+## Memories
+
+A memory is an abstract representation of a hardware memory. It is
+characterized by the following parameters.
+
+1.  A passive type representing the type of each element in the memory.
+
+2.  A positive integer representing the number of elements in the
+    memory.
+
+3.  A variable number of named ports, each being a read port, a write
+    port, or readwrite port.
+
+4.  A non-negative integer indicating the read latency, which is the
+    number of cycles after setting the port's read address before the
+    corresponding element's value can be read from the port's data
+    field.
+
+5.  A positive integer indicating the write latency, which is the number
+    of cycles after setting the port's write address and data before the
+    corresponding element within the memory holds the new value.
+
+6.  A read-under-write flag indicating the behaviour when a memory
+    location is written to while a read to that location is in progress.
+
+The following example demonstrates instantiating a memory containing 256
+complex numbers, each with 16-bit signed integer fields for its real and
+imaginary components. It has two read ports, `r1` and `r2`, and one
+write port, `w`. It is combinationally read (read latency is zero
+cycles) and has a write latency of one cycle. Finally, its
+read-under-write behavior is undefined.
+
+    mem mymem :
+      data-type => {real:SInt<16>, imag:SInt<16>}
+      depth => 256
+      reader => r1
+      reader => r2
+      writer => w
+      read-latency => 0
+      write-latency => 1
+      read-under-write => undefined
+
+In the example above, the type of `mymem` is:
+
+    {flip r1: {flip data: {real:SInt<16>, imag:SInt<16>},
+               addr: UInt<8>,
+               en: UInt<1>,
+               clk: Clock}
+     flip r2: {flip data: {real:SInt<16>, imag:SInt<16>},
+               addr: UInt<8>,
+               en: UInt<1>,
+               clk: Clock}
+     flip w: {data: {real:SInt<16>, imag:SInt<16>},
+              mask: {real:UInt<1>, imag:UInt<1>},
+              addr: UInt<8>,
+              en: UInt<1>,
+              clk: Clock}}
+
+The following sections describe how a memory's field types are
+calculated and the behavior of each type of memory port.
+
+### Read Ports
+
+If a memory is declared with element type `T`, has a size less than or
+equal to $2^N$, then its read ports have type:
```

Comment:
This doesn't render.

Reply:
Good catch. This is fixed in later commits.